### PR TITLE
[BUGFIX] Use one inspector for the same engine/execution_engine

### LINKED ISSUE #5933: [BUG] Snowflake metadata is not caching
My environment:  MacOS
I am trying to connect snowflake to my code with greatest expectations to get some results regarding data quality and validation.

Everything is working great except the fact that every second GE is sending this query to SF: 
(SHOW /* sqlalchemy:_get_schema_primary_keys */PRIMARY KEYS IN SCHEMA myschema), and also this query 

(SELECT … FROM information_schema.columns ic WHERE ic.table_schema=myschema)

Also, I tried to add cache_column_metadata=True in my connection string but it didn't solve the issue.
I saw this issue as well https://github.com/great-expectations/great_expectations/issues/3429 but the solution proposed is not working in my side (the proposed solution is to using ConfiguredAssetSqlDataConnector).

```
data_context_config = DataContextConfig(
    datasources={
        "my_snowflake_datasource": DatasourceConfig(
            class_name="Datasource",
            execution_engine={
        "class_name": "SqlAlchemyExecutionEngine",
"connection_string": connstring
            },
            data_connectors={
    "default_configured_data_connector_name": {
        "class_name": "ConfiguredAssetSqlDataConnector",
        "module_name": "great_expectations.datasource.data_connector",
         "assets": {
                "table": {
                    "splitter_method" : "_split_on_column_value",
                    "splitter_kwargs": {
                        "column_name": "ds"
                    }
                }
           },
    },
            
            }
        )
    },
    store_backend_defaults=FilesystemStoreBackendDefaults(root_directory=temp_dir.name)
)
```

```
context = BaseDataContext(project_config=data_context_config)


batch_request=BatchRequest(
    datasource_name="my_snowflake_datasource",
    data_connector_name="default_configured_data_connector_name",
    data_asset_name="table",
    data_connector_query={
        "batch_filter_parameters": {"ds": "2022-01-01"}
    },
    batch_spec_passthrough={"create_temp_table": False, 'use_quoted_name': True},
)

```

### PR DESCRIPTION
**Pull request by TrangPham**
Previously, we were getting a new inspector each time that "table.column_types" metric was being computed. This means that SQLAlchemy would not see that information as cached as it is for a different inspector instance.

This changes adds _inspector to the ExecutionEngine and uses that same inspector later on (making it a singleton-ish).

Some additional refactoring can be done to move the functionality of "get_column_metadata" into the different engines and just ask the engine for that information. This would mean a refactor of the functions that "table_column_types.py" uses.

Closes: #5933 

- [x] Description of PR changes above includes a link to [an existing GitHub issue](https://github.com/great-expectations/great_expectations/issues)
- [x] PR title is prefixed with one of: [BUGFIX], [FEATURE], [DOCS], [MAINTENANCE], [CONTRIB]
- [x] Code is linted - run `invoke lint` (uses `black` + `ruff`)
- [x] Appropriate tests and docs have been updated

For more information about contributing, see [Contribute](https://docs.greatexpectations.io/docs/contributing/contributing_checklist).

After you submit your PR, keep the page open and **monitor the statuses of the various checks made by our continuous integration process at the bottom of the page. Please fix any issues that come up** and [reach out on Slack](https://greatexpectations.io/slack) if you need help. Thanks for contributing!

### CODE REVIEW

File: great_expectations/expectations/metrics/table_metrics/table_column_types.py [hunk 1]
```diff
@@ -87,7 +87,9 @@ def _spark(  # noqa: PLR0913
         )
 
 
-def _get_sqlalchemy_column_metadata(engine, batch_data: SqlAlchemyBatchData):
+def _get_sqlalchemy_column_metadata(
+    execution_engine: SqlAlchemyExecutionEngine, batch_data: SqlAlchemyBatchData
+):
```
Comment (Kilo59) [inline great_expectations/expectations/metrics/table_metrics/table_column_types.py:95]: Sorry for the late review.

@TrangPham are you sure this is meant to be the `SqlAlchemyExecutionEngine` and not just the native Sqlalchemy `sqlachemy.engine,Engine` type?

A type annotation here would have been helpful 😄

File: tests/expectations/metrics/test_table_column_types.py [hunk 2]
```diff
@@ -35,3 +36,116 @@ def test_table_column_introspection(sa):
...
+    )
+
+    std, err = capfd.readouterr()
+
+    assert not std
+    assert not err
+
+
+@pytest.mark.sqlite
+def test_table_column_type__sqlalchemy_happy_path(sa, capfd):
```
Comment (Shinnnyshinshin) [inline tests/expectations/metrics/test_table_column_types.py]: Curious : were you able to get this test passing locally?

File: great_expectations/expectations/metrics/table_metrics/table_column_types.py (hunk 1 above)
Comment (TrangPham) [reply great_expectations/expectations/metrics/table_metrics/table_column_types.py:95]: Yes this is correct. It is being called in one place in this function which passes in the execution engine:

 https://github.com/great-expectations/great_expectations/blob/4c1302642ea559162f6a48bb45861cb5862c0cfc/great_expectations/expectations/metrics/table_metrics/table_column_types.py#L46

File: tests/expectations/metrics/test_table_column_types.py (hunk 2 above)
Comment (TrangPham) [reply tests/expectations/metrics/test_table_column_types.py]: Yes this passes locally for me

**Comment by TrangPham** [comment]
Updated the tests so it would work.

Seems like it passed for me locally, but within the github ci the sqlite file was being closed and capsys.readerrout().err was showing that the file was closed when trying to run empty SQL. This is for the happy path with used our metric resolution directly vs the other test which calls get_column without going through the metric resolution.

File: tests/expectations/metrics/test_table_column_types.py [hunk 3]
```diff
@@ -35,3 +51,136 @@ def test_table_column_introspection(sa):
...
+        assert results[("table.columns", (), ())] == [
+            "index",
+            "id",
+            "event_type",
+            "favorite_color",
+        ]
+
+        # Should have to re-inspect due to new batch
+        # But should use same inspector
+        assert mock_inspect.call_count == 1
```
Comment (Shinnnyshinshin) [inline tests/expectations/metrics/test_table_column_types.py:158]: 🚀

**Review by Shinnnyshinshin — APPROVED** [review]
LGTM! Thank you for this @TrangPham

**Pull request merged**